### ISSUE TITLE
Error when find an entity with undefined value

### ISSUE THREAD

**Issue by godjirax** [sebelga/gstore-node#176]
When i execute the findOne function on a Model, with a property and undefined value, an error is thrown : `Error: Unsupported field value, undefined, was provided.`

```js
MyModel.findOne({ myProperty: undefined })
  .then(console.log)
  .catch(error => {
    // Can't catch this
    console.error(error)
  })

```

Problem is: i can't catch this error and it crashes my process.

How can i handle this error ?

**Comment by lesmo** [comment]
This is an error on your implementation. That exception is thrown by the underlying [@google-cloud/datastore](https://github.com/googleapis/nodejs-datastore) library.

Exceptions are different to promise rejections. If you're trying to find entities whose ```myProperty``` is not set, you should instead use ```null```.

Your code would work if you do this:
```js
try {
  MyModel.findOne({ myProperty: undefined })
    .then(console.log)
    .catch(error => {
      // Promise rejection
      console.error(error)
    })
} catch (ex) {
  // Exception
  console.error(ex);
}
```

**Comment by godjirax** [comment]
We could use null, but the point is not there.

I'm not trying to find entities with a property not set.
I'm writing test cases and we were expecting an error to be thrown.
That's the case but it's not catchable.

In my point of view, it's not something desirable. If any error occurs, we want to handle it.

By the way, I think you wanted to write null instead of undefined: 

```js
try {
  MyModel.findOne({ myProperty: null })
    .then(console.log)
    .catch(error => {
      // Promise rejection
      console.error(error)
    })
} catch (ex) {
  // Exception
  console.error(ex);
}
```

I tried to patch this.
However I don't know if the responsibility falls on your library or Google's.

**Comment by lesmo** [comment]
Yeah, I meant to write ```undefined``` to try and workaround your code. The exception is thrown outside a promise, which is before control is returned to your code... so it can be handled in a ```try..catch```, but not through ```.catch```.

The exception is thrown in [@google-cloud/datastore/src/entity.ts:497](https://github.com/googleapis/nodejs-datastore/blob/b36c90f73342d43deb8be0a75ea4fda76a41afc9/src/entity.ts#L497). I'd need to check the callstack, but it's probably thrown when the filter method [is called by gstore-node](https://github.com/sebelga/gstore-node/blob/72fff67b064e2498b9081d3377f3bc6b0a76109c/lib/query.js#L126). As you can see in the code, the promise has not been created and thus the exception is not handled as a promise rejection.

It should be easy to patch by wrapping the entire ```findOne``` method body in a Promise. However I think this should be handled by the implementer, not the libraries. But that's just me.

I think a nicer way would be to just do:
```js
try {
  const entity = await MyModel.findOne({.myProperty: undefined });
  console.log(entity);
} catch (e) {
  console.log(e);
}
```
Which is currently how we're mostly handling stuff in my project. The ```try..catch``` handles both Promise rejections and exceptions thrown.

**Comment by godjirax** [comment]
I totally agree with you, this kind of error has to be handled by the implementer.

But with your snippet, `console.log(e)` is not called and the program crashes.

Therefore, the problem must be in the Google library.

After investigation, the error occurs in `@google-cloud/datastore/build/src/request.js` l.634. at `makeRequest(query);` call.
I think the error has to be handled here, with something like:

```js
stream.once('reading', () => {
  try {
    makeRequest(query);
  } catch(err) {
    stream.emit('error', err);
  }
});
```

After adding this try/catch, console.log in your snippet is properly called.

I'm going to open an issue on Google's repo.

**Comment by lesmo** [comment]
That's a nice find! If you open that issue, could you link it here for followup?

**Comment by sebelga** [comment]
Thanks for reporting and your finding @godjirax ! 
Could you add the link to the issue you opened in the Google repo and close the issue here once you did it?
Cheers! 👍

**Comment by godjirax** [comment]
I'm confused.

I wrote a piece of code to show the bug with native datastore driver, and i'm able to catch the error...
Can problem be in gstore-node library?

```js
const { Datastore } = require('@google-cloud/datastore');

const projectId = "projectId"
const namespace = "namespace"

const datastore = new Datastore({ projectId, namespace })

const query = datastore
  .createQuery('MyModel')
  .filter('property', undefined)

query.run()
  .then(results => {
    console.log('results', results)
  })
  .catch(error => {
    console.error('Catchable error?', error)
  })
```

**Comment by sebelga** [comment]
Hello,

I am not sure you can do a filter comparison against `undefined` afak. You must provide a value.
Have you tried to do it directly with the Datastore client?

I don't know your use case, but instead of searching for a property that hasn't been defined, why don't you use a flag (a boolean) that can either be `true` or `false`?

**Comment by godjirax** [comment]
I understand I musn't provide undefined value but something else.

Use case is: program creates a query from an object, giving a property value which can be undefined.
I can test property existence before, but for me if something bad happens in a library, I have to be able to catch all errors.

My bad about my last code. It shows the catchable error because I fixed the Google datastore library.

I created an issue on Google repo: https://github.com/googleapis/nodejs-datastore/issues/491
